latte/utils/ksg_estimator.py

```diff
@@ -0,0 +1,216 @@
+    knn.fit(X)
+    d, _ = knn.kneighbors(X)  # the first nearest neighbor is itself
+    return d[:, -1]  # returns the distance to the kth nearest neighbor
+
+
+def covar_to_corr(C):
+    assert np.allclose(C, C.T), "Covariance matrix not symmetric"
+    d = 1 / np.sqrt(np.diag(C))
+    # same as np.diag(d) @ C @ np.diag(d), but using broadcasting
+    return d * (d * C).T
+
+
+def entropy_gaussian(C):
+    """
+    Entropy of a gaussian variable with covariance matrix C
+    """
+    # Remember Covariance is dimensional variable. Corr is dimensionless.
+    # You can never take the log of a dimensional variable.
+    if np.isscalar(C):  # corr is just 1
+        return 0.5 * (1 + np.log(2 * pi))  # + 0.5 * np.log(1)
+    else:
+        corr = covar_to_corr(C)
+        n = corr.shape[0]  # dimension
+        return 0.5 * n * (1 + np.log(2 * pi)) + 0.5 * np.log(abs(det(corr)))
+
+
+def entropy(X, k=1, transform=DEFAULT_TRANFORM):
+    """Returns the approximate entropy of X via some knn estimator.
+
+    Parameters
+    ===========
+
+    X : array-like, shape (n_samples, n_features)
+        The data the entropy of which is computed
+
+    k : int, optional
+        number of nearest neighbors for density estimation
+
+    Notes
+    ======
+
+    Kozachenko, L. F. & Leonenko, N. N. 1987 Sample estimate of entropy
+    of a random vector. Probl. Inf. Transm. 23, 95-101.
+    See also: Evans, D. 2008 A computationally efficient estimator for
+    mutual information, Proc. R. Soc. A 464 (2093), 1203-1215.
+    and:
+    Kraskov A, Stogbauer H, Grassberger P. (2004). Estimating mutual
+    information. Phys Rev E 69(6 Pt 2):066138.
+    """
+    check = np.unique(X, axis=0)
+    if check.shape[0] == 1:
+        # deterministic variable has entropy 0
+        return 0.0
+    X = handle_transform(X, transform)
+
+    # Distance to kth nearest neighbor
+    R = nearest_distances(X, k)  # squared distances
+    N, D = X.shape
+    """
+    F. Perez-Cruz, (2008). Estimation of Information Theoretic Measures
+    for Continuous Random Variables. Advances in Neural Information
+    Processing Systems 21 (NIPS). Vancouver (Canada), December.
+
+    See eq (9):
+
+        (9) hhat = - mean(log(phat))
+        (3) phat = k / (n - 1) / volume_unit_ball / r ** d
+
+    So phat = d * mean(log(r)) + log(vub) + log(n - 1) - log(k)
+    In eq (20) of Kraskov (2003), it is different:
+
+        (20) hhat = -digamma(k) + digamma(n) + log(vub) + d / n * mean(log(2 * r))
+
+    See https://hal.inria.fr/hal-01272527/document probably for best description.
+
+    I think the confusion is that in the L_infinity norm, unitball is 2 ** D.
+
+    See also the dit project on github. Has a similar form.
+    """
+    R = R + np.finfo(X.dtype).eps
+    volume_unit_ball = (pi ** (D / 2)) / gamma(1 + D / 2)
+    return digamma(N) - digamma(k) + np.log(volume_unit_ball) + D * np.mean(np.log(R))
+
+
+def mutual_information(variables, k=1, transform=DEFAULT_TRANFORM):
```

Comment:
Out of curiosity, how close are these estimates to MINE?

This formula can be improved (individual errors on entropy don't cancel when summed/subtracted), but I guess it should be already quite good.